### ISSUE TITLE
HTTPAuthSessionWrapper incorrectly returns a 500 if the wrapped resource doesn't implement the requested method

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3932]
|<img alt="esteve's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| esteve reported|
|-|-|
|Trac ID|trac#3932|
|Type|defect|
|Created|2009-07-27 18:01:39Z|
|Branch|https://github.com/twisted/twisted/tree/httpauth-method-not-allowed-3932|

As the summary says, if a resource is wrapped using HTTPAuthSessionWrapper and an unimplemented method is requested by a client, a 500 response code (Invalid error) is returned instead of 405 (Method Not Allowed). An un-wrapped resource correctly returns a 405 response code.

Attachments:

* [webguard_405.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/6d7/6d77626b609e576e85e89fe52d4455cc5c3e3f1f/c66017ae293430cac2b60088c1ce20888596e775.py) (1033 bytes) - added by esteve on 2009-07-27 18:09:57Z - 
* [3932-fix-method-not-allowed-in-httpauthsessionwrapper.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/6d7/6d77626b609e576e85e89fe52d4455cc5c3e3f1f/e1377ea0bc1bb60fdb3ae7a6341408b032b2b03f.patch) (8740 bytes) - added by esteve on 2009-07-27 18:15:51Z - 
* [3932.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/6d7/6d77626b609e576e85e89fe52d4455cc5c3e3f1f/38ca8c7c469fa04684e5c1e6adbf87b9f1c0a197.diff) (693 bytes) - added by therve on 2009-08-16 23:59:26Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3932 3932
type__defect defect
reporter__esteve esteve
priority__normal normal
milestone__ 
branch__branches_httpauth_method_not_allowed_3932 branches/httpauth-method-not-allowed-3932
branch_author__therve__exarkun therve, exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1248717699000000 1248717699000000
changetime__1258930110000000 1258930110000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="esteve's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|esteve commented|
|-|-|

Given an example like one I attached (webguard_405.py), this is the traceback I get:


```
2009-07-27 20:07:31+0200 [-] Log opened.
2009-07-27 20:07:31+0200 [-] twisted.web.server.Site starting on 8889
2009-07-27 20:07:31+0200 [-] Starting factory <twisted.web.server.Site instance at 0xd2eb00>
2009-07-27 20:07:33+0200 [HTTPChannel,0,127.0.0.1] 127.0.0.1 - - [27/Jul/2009:18:07:32 +0000] "GET / HTTP/1.0" 401 12 "-" "Wget/1.10.2"
2009-07-27 20:07:33+0200 [HTTPChannel,1,127.0.0.1] Unhandled Error
        Traceback (most recent call last):
          File "/usr/lib/python2.5/site-packages/twisted/web/server.py", line 157, in render
            body = resrc.render(self)
          File "/usr/lib/python2.5/site-packages/twisted/web/util.py", line 94, in render
            self.d.addCallback(self._cbChild, request).addErrback(
          File "/usr/lib/python2.5/site-packages/twisted/internet/defer.py", line 195, in addCallback
            callbackKeywords=kw)
          File "/usr/lib/python2.5/site-packages/twisted/internet/defer.py", line 186, in addCallbacks
            self._runCallbacks()
        --- <exception caught here> ---
          File "/usr/lib/python2.5/site-packages/twisted/internet/defer.py", line 328, in _runCallbacks
            self.result = callback(self.result, *args, **kw)
          File "/usr/lib/python2.5/site-packages/twisted/web/util.py", line 100, in _cbChild
            result = resource.getChildForRequest(child, request).render(request)
          File "/usr/lib/python2.5/site-packages/twisted/web/_auth/wrapper.py", line 171, in render
            return super(ResourceWrapper, self).render(request)
          File "/usr/lib/python2.5/site-packages/twisted/web/resource.py", line 189, in render
            raise UnsupportedMethod(getattr(self, 'allowedMethods', ()))
        twisted.web.server.UnsupportedMethod: ()
```

**Comment by twisted-trac** [comment]
|<img alt="esteve's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|esteve removed owner|
|-|-|

Here's an attempt at fixing this issue, it includes a test in test_httpauth.py and all the tests for twisted.web pass. It changes the code in the DeferredResource class to trap UnsupportedMethod exceptions (raised by Resource#render in case the requested method doesn't exist) in an errback and send the appropriate response code and contents.

I'm not particularly happy with the way I did it, as I had to move out some code for Request#render to two separate functions and I'm not sure if the test belongs in the HTTPAuthHeaderTests testcase. Anyway, feel free to criticize it mercilessly :-)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve set owner to esteve|
|-|-|

I'd like to propose a different approach, with the latest patch: do you think it would solve your problem?

**Comment by twisted-trac** [comment]
|<img alt="esteve's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|esteve set status to `assigned`|
|-|-|

Replying to [therve](#note_3):
> I'd like to propose a different approach, with the latest patch: do you think it would solve your problem?

Your patch is way cleaner than mine, however the tests don't pass, because DummyRequest doesn't implement render(). I ran into that problem before submitting my patch, so that's why I chose to break Resource#render into smaller methods, but I guess the cleanest approach would be to apply your patch and implement DummyRequest#render

**Comment by twisted-trac** [comment]
|<img alt="esteve's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|esteve commented|
|-|-|

Replying to [esteve](#note_4):
> patch, so that's why I chose to break Resource#render into smaller methods, but I 

I meant twisted.web.server.Request, not Resource

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [27563]) Branching to 'httpauth-method-not-allowed-3932'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [27564]) Apply therve's 3932.diff

refs [#3932](https://github.com/twisted/twisted/issues/3932)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

So I did a few things.

I made some minor cleanups to the `util.py` module, nothing too scary.

I moved the `test_httpauth.render` free-function onto `DummyRequest` as its render method and tried to capture some of my thoughts about it in its docstring.

I added a test for `DeferredResource` that verifies that it uses the request object to render the resulting resource.  Since `DeferredResource` is where this bug was actually present, this was the best test I could think of.  Testing here was a little bit of a challenge, since the bug was fixed by deleting a bunch of custom code and only adding back a little bit of code to integrate with some other code which already worked properly (and has some tests).  Short of running the unit tests against a real `Request`, I'm not sure what else to do (it'd be neat if we could do that someday though).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set owner to @exarkun|
|-|-|

Looks good, +1.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `closed`|
|-|-|

(In [27595]) Merge httpauth-method-not-allowed-3932

Author: therve, exarkun
Reviewer: therve
Fixes: [#3932](https://github.com/twisted/twisted/issues/3932)

Change `DeferredResource` so that it uses the request object to call the
`render` method of whatever resource its deferred eventually fires with.
This makes handling of the result (by return or exception) of such render
methods consistent with normal handling.  Previously only strings and
`NOT_DONE_YET` were correctly handled by `DeferredResource`, nothing else
(for example raising `UnsupportedMethod` was not handling, resulting in
a 500 response instead of the correct 405 or 501).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
